chatkit-core/src/main/kotlin/com/pusher/chatkit/subscription/SubscriptionListener.kt

```diff
@@ -0,0 +1,30 @@
+package com.pusher.chatkit.subscription
+
+import com.pusher.platform.SubscriptionListeners
+import elements.EOSEvent
+import elements.Headers
+import elements.SubscriptionEvent
+
+/**
+ * Platform SubscriptionListeners, but as an interface.
+ *
+ * In future, consider pushing this interface down in to platform as an alternative to the
+ * SubscriptionListeners data-class-full-of-closures approach.
+ */
+internal interface SubscriptionListener<A> {
+    fun onEnd(error: EOSEvent?)
+    fun onError(error: elements.Error)
+    fun onEvent(elementsEvent: SubscriptionEvent<A>)
+    fun onOpen(headers: Headers)
+    fun onRetrying()
+    fun onSubscribe()
```

Comment:
```suggestion
    fun onSubscribe() = { }
```
and for other unlikely used callbacks (removing the need of empty implementations)

Reply:
I was planning on implementing them all...

Reply:
oh really, well then please don't add the default empty implementation if they all will for sure need to be implemented in every class declaring this interface

Reply:
Cool. If any are not required, I'll drop them from the interface towards the end of the work.